gprofiler/merge.py

```diff
@@ -148,32 +148,27 @@ def add_highest_avg_depth_stacks_per_process(
         if fp_frame_count_average > dwarf_frame_count_average:
             merged_pid_to_stacks_counters[pid] = fp_collapsed_stacks_counters
         else:
-            dwarf_collapsed_stacks_counters = scale_dwarf_samples_count(
+            dwarf_collapsed_stacks_counters = scale_sample_counts(
                 dwarf_collapsed_stacks_counters, fp_to_dwarf_sample_ratio
             )
             merged_pid_to_stacks_counters[pid] = dwarf_collapsed_stacks_counters
 
 
-def scale_dwarf_samples_count(
-    dwarf_collapsed_stacks_counters: StackToSampleCount,
-    fp_to_dwarf_sample_ratio: float,
-) -> StackToSampleCount:
-    if fp_to_dwarf_sample_ratio == 1:
-        return dwarf_collapsed_stacks_counters
-    # scale the dwarf stacks to the FP stacks to avoid skewing the results
-    for stack, sample_count in dwarf_collapsed_stacks_counters.items():
-        new_count = sample_count * fp_to_dwarf_sample_ratio
+def scale_sample_counts(stacks: StackToSampleCount, ratio: float) -> StackToSampleCount:
+    if ratio == 1:
+        return stacks
+
+    scaled_stacks: StackToSampleCount = StackToSampleCount()
+    for stack, count in stacks.items():
+        new_count = count * ratio
         # If we were to round all of the sample counts it could skew the results. By using a random factor,
         # we mostly solve this by randomly rounding up / down stacks.
         # The higher the fractional part of the new count, the more likely it is to be rounded up instead of down
-        new_count = math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)
-        if new_count == 0:
-            # TODO: For more accurate truncation, check if there's a common frame for the truncated stacks and combine
-            #  them
-            continue
-        dwarf_collapsed_stacks_counters[stack] = new_count
-    # Note - returning the value is not necessary, but is done for readability
-    return dwarf_collapsed_stacks_counters
+        scaled_value = math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)
```

Comment:
Umm... I'm either missing the point here, or, this `if` is swapped? For `new_count` with "large" fractions (e.g `.95`) this expression is more likely to pick the `floor` path instead of the `ceil` path. (and vice-versa for small fractions):

```
In [27]: new_count                                                                                                                                                                                         
Out[27]: 3.99

In [28]: math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)                                                                                                  
Out[28]: 3

In [29]: math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)                                                                                                  
Out[29]: 3

In [32]: new_count = 2.01                                                                                                                                                                                  

In [33]: math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)                                                                                                  
Out[33]: 3

In [34]: math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)                                                                                                  
Out[34]: 3

In [35]: math.ceil(new_count) if new_count - int(new_count) <= random.random() else math.floor(new_count)                                                                                                  
Out[35]: 3
```

I suppose the condition needs to be swapped.

Reply:
Yea, should be `random <= fraction`
This logic existed before the PR for dwarf scaling, fixed now

Reply:
Mhm, was broken before.